libraries/AP_HAL_Linux/CAN.cpp

```diff
@@ -118,26 +118,18 @@ int LinuxCAN::openSocket(const std::string& iface_name)
 {
     errno = 0;
 
-    const int s = socket(PF_CAN, SOCK_RAW, CAN_RAW);
+    int s = socket(PF_CAN, SOCK_RAW, CAN_RAW);
     if (s < 0) {
         return s;
     }
 
-    class RaiiCloser
-    {
-        int fd_;
-    public:
-        RaiiCloser(int filedesc) : fd_(filedesc)
-        { }
-        ~RaiiCloser()
-        {
-            if (fd_ >= 0) {
-                hal.console->printf("SocketCAN: closing fd\n");
-                close(fd_);
-            }
+    std::shared_ptr<int> fd_guard(new int(s), [](const int *fd) {
```

Comment:
oh... no, that's not what I meant. you shouldn't need to allocate the int here. See the example I posted or how it's done on mavlink-router: https://github.com/intel/mavlink-router/blob/master/src/mavlink-router/logendpoint.cpp#L105.


Or simply don't do it... I see little value in it here and just make the code more complex.

Reply:
that's why I told to assign to `= -1` in the end, so the fd would not be closed.

The reality is that C++ sucks here and it's too verbose to do what a simple `__attribute__((cleanup))` does in C. It also more powerful, though, but then it should be generic enough to be adapted to other places in order to be justifiable.

Reply:
@lucasdemarchi can you help out and be explicit with what you'd like it to be? Either post the exact code change here or on a commit on your own branch that we can cherrypick?

Reply:
Yes, so, therefore, I cannot do this just as in the link you gave me.

Reply:
@Programmer74 would be simpler to just assign the var to -1, as I suggested, and return another var. 

```c++
std::shared_ptr<void> defer(s, [](int fd) { if (fd >= 0) close(fd); });
const int ret = s;

...

s = -1;
return ret;
```

Or... just remove this commit and leave as is, then when there's a generic solution I can replace this. Because right now you are unnecessarily allocating one integer ... and leaking it in case fd is negative.

What I was thinking was something generic like this: https://codereview.qt-project.org/#/c/206467/6//ALL,unified  which is very like the "raiicloser" here, but in a way that resembles the `defer` from go and that is not specialized to a single use-case.

Reply:
@lucasdemarchi Changed the ```defer``` as in your example and rebased everything.